test/integration/main.cpp

```diff
@@ -2094,4 +2103,197 @@ TEST_F(ZNCTest, ModuleCrypt) {
     Z;
 }
 
+#ifdef WITH_OTR
+static void
+IsOtrCiphertext(std::string line) {
+    EXPECT_THAT(line, MatchesRegex(":?\\S* ?PRIVMSG \\S+ :.OTR.*\\.?,?\\r?"));
+}
+
+TEST_F(ZNCTest, ModuleOtr) {
+    QFile conf(m_dir.path() + "/configs/znc.conf");
+    ASSERT_TRUE(conf.open(QIODevice::Append | QIODevice::Text));
+    QTextStream(&conf) << "ServerThrottle = 1\n";
+    auto znc = Run();
+    Z;
+    auto ircd1 = ConnectIRCd();
+    Z;
+    auto client1 = LoginClient();
+    Z;
+    client1.Write("znc loadmod controlpanel");
+    client1.Write("PRIVMSG *controlpanel :CloneUser user user2");
+    client1.ReadUntil("User [user2] added!");
+    client1.Write("PRIVMSG *controlpanel :Set Nick user2 nick2");
+    Z;
+    client1.Write("znc loadmod otr");
+    client1.ReadUntil("Loaded module");
+    Z;
+    auto ircd2 = ConnectIRCd();
+    Z;
+    auto client2 = ConnectClient();
+    client2.Write("PASS user2:hunter2");
+    client2.Write("NICK nick2");
+    client2.Write("USER user2/test x x :x");
+    Z;
+    client2.Write("znc loadmod otr");
+    client2.ReadUntil("Loaded module");
+    Z;
+
+    // genkey without --really should produce a warning
+    client1.Write("PRIVMSG *otr :genkey");
+    client1.ReadUntil("genkey --really");
+    Z;
+
+    // generate keys
+    client1.Write("PRIVMSG *otr :genkey --really");
+    client1.ReadUntil("Key generation finished.");
+    Z;
+    client2.Write("PRIVMSG *otr :genkey --really");
+    client2.ReadUntil("Key generation finished.");
+    Z;
+
+    // do not overwrite existing key without --overwrite
+    client2.Write("PRIVMSG *otr :genkey --really");
+    client2.ReadUntil("genkey --overwrite");
+    Z;
+
+    // OTR handshake
+    client1.Write("PRIVMSG nick2 :?OTR?");
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    Z;
+    ircd2.RelayPrivMsg(ircd1, ":nick2!user2@user2/test ", IsOtrCiphertext);
+    Z;
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    Z;
+    ircd2.RelayPrivMsg(ircd1, ":nick2!user2@user2/test ", IsOtrCiphertext);
+    Z;
+    ircd2.RelayPrivMsg(ircd1, ":nick2!user2@user2/test ", IsOtrCiphertext);
+    Z;
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    Z;
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
```

Comment:
Yeah ... couldn't think of less magical way to let the two znc instances exchange messages until the handshake is done.